### DISCUSSION TITLE
Manage native (C++) build dependencies

### DISCUSSION THREAD

**Discussion by rainman110** [scikit-build/scikit-build#645] [Q&A]
What are the best practices, to install and use non-python dependencies with skbuild, in particular with cmake?

Is there any good example, that requires e.g. the boost libraries?

Ideally, I'd like to combine 
  - *skbuild* for the build and packaging process and 
  - *conan* to install dependencies

Is it possible to call `conan install` automatically by setup.py from within cmake's build directory?

**Comment by zouhairm** [comment]
@rainman110 - I'm curious how you've ended up handling this?

**Reply by rainman110** [reply]
I did not find a "best" solution unfortunately. My workaround is to manually set the "CMAKE_PREFIX_PATH" environment variable to all external dependencies first and then build the project (via scikit build).

Another workaround is to integrate the dependencies as subprojects / submodules and build everything together.

**Comment by yssource** [comment]
I'm also curious about how to make scikit-build work with [conan package manager](https://docs.conan.io/en/latest/)?

**Comment by zouhairm** [comment]
with the help of some cmake gurus, I ended up implementing this by relying on dependency providers. See details below:

 
- added a "package provider" file called `conan_pkg_provide_dependency.cmake.j2`
(note, this is a jinja template as it was part of some more complicated code, but the {{ }} should be obvious
```
set(CONAN_PKG_PROVIDER_INSTALL_DIR ${CMAKE_BINARY_DIR}/conan_packages
  CACHE PATH "The directory this provider installs packages to"
)

macro(conan_pkg_provide_dependency method cmake_package)
  if(${cmake_package} MATCHES "{{ target_name }}")
      execute_process(
        COMMAND conan install {{ target_conan_name_version_user_channel }} --build=missing -g deploy -if ${CONAN_PKG_PROVIDER_INSTALL_DIR}
        COMMAND_ERROR_IS_FATAL ANY
      )
      find_package(${cmake_package} PATHS ${CONAN_PKG_PROVIDER_INSTALL_DIR}/{{ target_conan_name }} NO_DEFAULT_CMAKE_PATH BYPASS_PROVIDER)
  endif()
endmacro()

cmake_language(
  SET_DEPENDENCY_PROVIDER conan_pkg_provide_dependency
  SUPPORTED_METHODS FIND_PACKAGE
)
```
See https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html

- in my CMakeLists.txt I added:
```
find_package({{ target_name }} REQUIRED)
```
which will end up using the package provider defined above since the target_name would match

- finally, in my setup() in the setup.py, I passed to cmake the following args:
```
cmake_args=[f'-DCMAKE_PROJECT_TOP_LEVEL_INCLUDES={Path(__file__).parent.absolute()}/conan_pkg_provide_dependency.cmake']
```
See https://cmake.org/cmake/help/latest/variable/CMAKE_PROJECT_TOP_LEVEL_INCLUDES.html

**Reply by rainman110** [reply]
That looks indeed like a working solution! Thanks for sharing this! 🙏 I am going to check this out.
